### DISCUSSION TITLE
Testing async initialization in AsyncNotifier?

### DISCUSSION THREAD

**Discussion by joakimunge** [rrousselGit/riverpod#2864] [Q&A] [answered]
Hi, 

I'm trying to write tests for an AsyncNotifier (AuthController in this case) which uses async initialization. I have read [this](https://codewithandrea.com/articles/unit-test-async-notifier-riverpod/) article by Andrea which was a big help for starters, but his examples do not cover async initialization _(which is a shame since that is one of the biggest draws of this Notifier class)._

The issue I'm having is that the `Listener` only gets called once with `AsyncLoading` and does not get called with the next `AsyncData` which in this case would be an `AuthState`.

For this to work I have to `await` something. It does not matter what or how long. Even `await Future<void>.value()` is enough for it to properly function. I just can not really understand why. Is there another way to properly test the async nature of the initialization?

Here is a minimal implementation of the AsyncNotifier:

```dart
class AuthController extends AsyncNotifier<AuthState> {
  @override
  FutureOr<AuthState> build() async {
    {..} // Some async ops
    return const AuthState(authStatus: AuthStatus.Unauthenticated);
  }
```

And here is the test file

```dart
class MockAuthService extends Mock implements AuthService {}

class Listener<T> extends Mock {
  void call(T? previous, T next);
}

void main() {
  ProviderContainer makeProviderContainer(MockAuthService authRepository) {
    final container = ProviderContainer(
      overrides: [
        authServiceProvider.overrideWithValue(authRepository),
      ],
    );
    addTearDown(container.dispose);
    return container;
  }

  group('initialization', () {
    test('initial state is AsyncData', () async {
      final authRepository = MockAuthService();
      final container = makeProviderContainer(authRepository);
      final listener = Listener<AsyncValue<AuthState>>();
      container.listen(
        authControllerProvider,
        listener,
        fireImmediately: true,
      );

      await Future<
          void>.value(); // Without awaiting something, the listener will only be called once (AsyncLoading)

      verifyInOrder([
        () => listener(
              null,
              const AsyncLoading<AuthState>(),
            ),
        () => listener(
              const AsyncLoading<AuthState>(),
              const AsyncData<AuthState>(AuthState(
                authStatus: AuthStatus.Unauthenticated,
              )),
            ),
      ]);
    });
  });
}
```

**Comment by rrousselGit** [comment, marked as answer]
Are you looking for `await container.read(authControllerProvider.future)`?

**Reply by joakimunge** [reply]
Indeed I think that is what I was looking for. That being said, do you have any idea as to why it would function with a simple 
`await Future<void>.value();` ? It seems so arbitrary, but I guess that has more to do with how asynchronization works in dart?

**Reply by rrousselGit** [reply]
Your `build` is `async`. So the state is initialized asynchronously (even if it does not include any await)

**Reply by joakimunge** [reply]
Right. So even if it's not actually doing any async ops in the test, it's still a Future so we need to have an `await` somewhere in the test, even if we do not await our future from the controller?

**Reply by rrousselGit** [reply]
Yes. You could remove the async keyword. In which case your provider _could_ be sync

**Reply by joakimunge** [reply]
Appreciate it @rrousselGit . Threw in a Future.delayed for a sanity check and indeed now I have to await the correct future. Learned something new.

Thanks!
